jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Callback.java

```diff
@@ -389,6 +391,153 @@ static Callback from(Callback callback1, Callback callback2)
         return combine(callback1, callback2);
     }
 
+    /**
+     * A combination of multiple Callbacks, that must all be completed before a specific callback is completed.
+     * For example: <pre>{@code
+     *   void sendToAll(String message, Collection<Channel> channels, Callback callback)
+     *   {
+     *       try (Callback.Combination combination = new Callback.Combination(callback))
+     *       {
+     *           for (Channel channel : channels)
+     *               channel.send(message, combination.newCallback());
+     *       }
+     *   }
+     * }</pre>
+     */
+    class Combination implements AutoCloseable
+    {
+        private final Callback andThen;
+        private final AtomicStampedReference<Throwable> state;
+
+        /**
+         * Create a new empty combined callback.
+         * @param andThen The {@code Callback} to complete once all {@code Callbacks} in the combination are complete
+         *                and this combination is {@link #close() closed}.
+         */
+        public Combination(Callback andThen)
+        {
+            this(andThen, null);
+        }
+
+        /**
+         * Create a new empty combined callback with a forced failure and invocation type.
+         *
+         * @param andThen The {@code Callback} to complete once all {@code Callbacks} in the combination are complete
+         *                and this combination is {@link #close() closed}.
+         * @param failure If not {@code null}, force a failure, so that the {@link Callback#failed(Throwable)} method
+         * will always be called on the {@code Callback} passed to the constructor, once all the
+         * combined callbacks are completed and the {@code Combination} is {@link #close() closed}.
+         */
+        public Combination(Callback andThen, Throwable failure)
+        {
+            this.andThen = andThen;
+            // initial stamp is -1 to indicate not closed and requiring 1 completion from close.
+            this.state = new AtomicStampedReference<>(failure, -1);
+        }
+
+        /**
+         * Create a new {@code Callback} as part of this combination.
+         *
+         * @return A {@code Callback} that must be completed before the callback passed to
+         *         the constructor is completed.
+         * @throws IllegalStateException if the combination has already been completed.
+         */
+        public Callback newCallback()
+        {
+            while (true)
+            {
+                int s = state.getStamp();
+                Throwable failure = state.getReference();
+
+                if (s >= 0)
+                    throw new IllegalStateException("closed");
+                // we can only create new callbacks when not closed, so we decrement to make a more negative stamp.
+                if (!state.compareAndSet(failure, failure, s--, s))
```

Comment:
Is this right? The third parameter is the expected stamp, so shouldn't be `s`, and the fourth parameter be `--s`?

If the code is wrong, it's not tested and should have a test.

Reply:
`s--, s` and `s, --s` are identical.  The decrement happens between the two uses of s

Reply:
Ugh, this is just obscure.
Please change it to:

```suggestion
                if (!state.compareAndSet(failure, failure, s, s - 1))
```

Reply:
really?  It is the basic meaning of `--s` vs `s--` ?  But OK,  I'll change it here